security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+// 1. does not have a good support on root cert reloading
+// 2. will ignore basic certificate check when setting InsecureSkipVerify to true
+func buildVerifyFunc(c *advancedTLSCreds,
+	serverName string,
+	rawConn net.Conn) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	verifyFunc := func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		var rootCAs *x509.CertPool
+		if c.isClient {
+			rootCAs = c.config.RootCAs
+		} else {
+			rootCAs = c.config.ClientCAs
+		}
+		// reload root CA certs
+		if rootCAs == nil && c.getRootCAs != nil {
+			reloadRootCAs, err := c.getRootCAs(rawConn, rawCerts)
+			if err != nil {
+				return err
+			}
+			rootCAs = reloadRootCAs
+		}
+		// verify peers' certificates against RootCAs and get verifiedChains
+		certs := make([]*x509.Certificate, len(rawCerts))
+		for i, asn1Data := range rawCerts {
+			cert, _ := x509.ParseCertificate(asn1Data)
+			certs[i] = cert
+		}
+		opts := x509.VerifyOptions{
+			Roots:         rootCAs,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+		}
+		if !c.isClient {
+			opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
+		}
+		for _, cert := range certs[1:] {
+			opts.Intermediates.AddCert(cert)
+		}
+		// We use default hostname check if users doesn't specify verifyFunc function
+		if c.isClient && c.verifyFunc == nil && serverName != "" {
+			opts.DNSName = serverName
+		}
+		verifiedChains, err := certs[0].Verify(opts)
+		if err != nil {
+			return err
+		}
+		if c.isClient && c.verifyFunc != nil {
+			return c.verifyFunc(serverName, rawCerts, verifiedChains)
+		}
+		return nil
+	}
+	return verifyFunc
+}
+
+const alpnProtoStrH2 = "h2"
+
+func appendH2ToNextProtos(ps []string) []string {
+	for _, p := range ps {
+		if p == alpnProtoStrH2 {
+			return ps
+		}
+	}
+	ret := make([]string, 0, len(ps)+1)
+	ret = append(ret, ps...)
+	return append(ret, alpnProtoStrH2)
+}
+
+// syscallConn keeps reference of rawConn to support syscall.Conn for channelz.
+// SyscallConn() (the method in interface syscall.Conn) is explicitly
+// implemented on this type,
+//
+// Interface syscall.Conn is implemented by most net.Conn implementations (e.g.
+// TCPConn, UnixConn), but is not part of net.Conn interface. So wrapper conns
+// that embed net.Conn don't implement syscall.Conn. (Side note: tls.Conn
+// doesn't embed net.Conn, so even if syscall.Conn is part of net.Conn, it won't
+// help here).
+type syscallConn struct {
+	net.Conn
+	// sysConn is a type alias of syscall.Conn. It's necessary because the name
+	// `Conn` collides with `net.Conn`.
+	sysConn
+}
+
+// WrapSyscallConn tries to wrap rawConn and newConn into a net.Conn that
+// implements syscall.Conn. rawConn will be used to support syscall, and newConn
+// will be used for read/write.
+//
+// This function returns newConn if rawConn doesn't implement syscall.Conn.
+func WrapSyscallConn(rawConn, newConn net.Conn) net.Conn {
```

Comment:
Can grpc-go expose this function to public, so that we don't have to copy the implementation?
@dfawley

Reply:
Yes, and another alternative might be we keep this function for now, and depend on the exposed one later, if it doesn't raise any other concerns.